File: app/src/main/java/com/kimjisub/launchpad/manager/ThemeResources.java
package com.kimjisub.launchpad.manager;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;

import com.kimjisub.launchpad.BuildConfig;
import com.kimjisub.launchpad.R;

public class ThemeResources {
	String packageName;

	Resources customRes;
	Resources defaultRes;

	public Drawable icon;
	public String name, description, author;
	public String version;

	public Drawable playbg, custom_logo;
	public Drawable btn, btn_;
	public Drawable chainled, chain, chain_, chain__;
	public Drawable phantom, phantom_;
	public Drawable xml_prev, xml_play, xml_pause, xml_next;
	public int checkbox, trace_log, option_window, option_window_checkbox, option_window_btn, option_window_btn_text;
	public boolean isChainLED = true;

	public ThemeResources(Context context, String packageName, boolean fullLoad) throws Exception {
		this.packageName = packageName;

		customRes = context.getPackageManager().getResourcesForApplication(packageName);
		defaultRes = context.getResources();

		////////////////////////////////////////////////////////////////////////////////////////////////

		icon = getCustomDrawable("theme_ic", R.drawable.theme_ic);

		version = context.getPackageManager().getPackageInfo(packageName, 0).versionName;

		name = getCustomString("theme_name", R.string.theme_name);
		description = getCustomString("theme_description", R.string.theme_description);
		author = getCustomString("theme_author", R.string.theme_author);

		if (!fullLoad)
			return;

		////////////////////////////////////////////////////////////////////////////////////////////////

		// Drawable
		playbg = getCustomDrawable("playbg", R.drawable.playbg);
		try {
			custom_logo = getCustomDrawable("custom_logo");
		} catch (Exception ignore) {
			custom_logo = null;
		}
		btn = getCustomDrawable("btn", R.drawable.btn);
		btn_ = getCustomDrawable("btn_", R.drawable.btn_);
		try {
			chainled = getCustomDrawable("chainled");
		} catch (Exception e) {
			isChainLED = false;
			chain = getCustomDrawable("chain");
			chain_ = getCustomDrawable("chain_");
			chain__ = getCustomDrawable("chain__");
		}
		phantom = getCustomDrawable("phantom");
		try {
			phantom_ = getCustomDrawable("phantom_");
		} catch (Exception ignore) {
		}
		xml_prev = getCustomDrawable("xml_prev", R.drawable.xml_prev);
		xml_play = getCustomDrawable("xml_play", R.drawable.xml_play);
		xml_pause = getCustomDrawable("xml_pause", R.drawable.xml_pause);
		xml_next = getCustomDrawable("xml_next", R.drawable.xml_next);


		// Color
		checkbox = getCustomColor("checkbox", R.color.checkbox);
		trace_log = getCustomColor("trace_log", R.color.trace_log);
		option_window = getCustomColor("option_window", R.color.option_window);
		option_window_checkbox = getCustomColor("option_window_checkbox", R.color.option_window_checkbox);
		option_window_btn = getCustomColor("option_window_btn", R.color.option_window_btn);
		option_window_btn_text = getCustomColor("option_window_btn_text", R.color.option_window_btn_text);
	}

	public ThemeResources(Context context, boolean fullLoad) throws Exception {
		defaultRes = context.getResources();

		////////////////////////////////////////////////////////////////////////////////////////////////

		icon = defaultRes.getDrawable(R.drawable.theme_ic);

		version = BuildConfig.VERSION_NAME;

		name = defaultRes.getString(R.string.theme_name);
		description = defaultRes.getString(R.string.theme_description);
		author = defaultRes.getString(R.string.theme_author);

		if (!fullLoad)
			return;

		////////////////////////////////////////////////////////////////////////////////////////////////

		// Drawable
		playbg = defaultRes.getDrawable(R.drawable.playbg);
		custom_logo = defaultRes.getDrawable(R.drawable.custom_logo);
		btn = defaultRes.getDrawable(R.drawable.btn);
		btn_ = defaultRes.getDrawable(R.drawable.btn_);
		chainled = defaultRes.getDrawable(R.drawable.chainled);
		phantom = defaultRes.getDrawable(R.drawable.phantom);
		phantom_ = defaultRes.getDrawable(R.drawable.phantom_);
		xml_prev = defaultRes.getDrawable(R.drawable.xml_prev);
		xml_play = defaultRes.getDrawable(R.drawable.xml_play);
		xml_pause = defaultRes.getDrawable(R.drawable.xml_pause);
		xml_next = defaultRes.getDrawable(R.drawable.xml_next);

		// Color
		checkbox = defaultRes.getColor(R.color.checkbox);
		trace_log = defaultRes.getColor(R.color.trace_log);
		option_window = defaultRes.getColor(R.color.option_window);
		option_window_checkbox = defaultRes.getColor(R.color.option_window_checkbox);
		option_window_btn = defaultRes.getColor(R.color.option_window_btn);
		option_window_btn_text = defaultRes.getColor(R.color.option_window_btn_text);
	}

	private Drawable getCustomDrawable(String customId) throws Exception {
		int resId = getResourceId("drawable", customId);
		return customRes.getDrawable(resId);
	}

	private int getCustomColor(String customId) throws Exception {
		int resId = getResourceId("color", customId);
		return customRes.getColor(resId);
	}

	private String getCustomString(String customId) throws Exception {
		int resId = getResourceId("string", customId);
		return customRes.getString(resId);
	}

	private Drawable getCustomDrawable(String customId, int defaultId) {
		try {
			return getCustomDrawable(customId);
		} catch (Exception e) {
			return defaultRes.getDrawable(defaultId);
		}
	}

	private int getCustomColor(String customId, int defaultId) {
		try {
			return getCustomColor(customId);
		} catch (Exception e) {
			return defaultRes.getColor(defaultId);
		}
	}

	private String getCustomString(String customId, int defaultId) {
		try {
			return getCustomString(customId);
		} catch (Exception e) {
			return defaultRes.getString(defaultId);
		}
	}

	private int getResourceId(String type, String customId) throws Exception {
		return customRes.getIdentifier(packageName + ":" + type + "/" + customId, null, null);
	}
}

File: app/src/main/java/com/kimjisub/launchpad/manager/ThemeResources.kt
package com.kimjisub.launchpad.manager

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.os.Build
import androidx.core.content.res.ResourcesCompat
import com.kimjisub.launchpad.BuildConfig
import com.kimjisub.launchpad.R.color
import com.kimjisub.launchpad.R.drawable
import com.kimjisub.launchpad.R.string

/**
 * 기본 테마: 앱 자체 리소스를 직접 사용 (PackageManager 불필요)
 */
class DefaultThemeResources(
	context: Context,
	fullLoad: Boolean = false,
) : IThemeResources {
	private val res: Resources = context.resources

	override var icon: Drawable = requireNotNull(ResourcesCompat.getDrawable(res, drawable.theme_ic, null))
	override var name: String = res.getString(string.theme_name)
	override var author: String = res.getString(string.theme_author)
	override var version: String = BuildConfig.VERSION_NAME

	override var playbg: Drawable? = null
	override var customLogo: Drawable? = null
	override var btn: Drawable? = null
	override var btnPressed: Drawable? = null
	override var chainled: Drawable? = null
	override var chain: Drawable? = null
	override var chainSelected: Drawable? = null
	override var chainGuide: Drawable? = null
	override var phantom: Drawable? = null
	override var phantomVariant: Drawable? = null
	override var xmlPrev: Drawable? = null
	override var xmlPlay: Drawable? = null
	override var xmlPause: Drawable? = null
	override var xmlNext: Drawable? = null
	override var checkbox: Int? = null
	override var traceLog: Int? = null
	override var optionWindow: Int? = null
	override var optionWindowCheckbox: Int? = null
	override var isChainLed = true

	init {
		if (fullLoad) {
			playbg = ResourcesCompat.getDrawable(res, drawable.playbg, null)
			btn = ResourcesCompat.getDrawable(res, drawable.btn, null)
			btnPressed = ResourcesCompat.getDrawable(res, drawable.btn_, null)
			chainled = ResourcesCompat.getDrawable(res, drawable.chainled, null)
			phantom = ResourcesCompat.getDrawable(res, drawable.phantom, null)
			xmlPrev = ResourcesCompat.getDrawable(res, drawable.xml_prev, null)
			xmlPlay = ResourcesCompat.getDrawable(res, drawable.xml_play, null)
			xmlPause = ResourcesCompat.getDrawable(res, drawable.xml_pause, null)
			xmlNext = ResourcesCompat.getDrawable(res, drawable.xml_next, null)
			checkbox = ResourcesCompat.getColor(res, color.checkbox, null)
			traceLog = ResourcesCompat.getColor(res, color.trace_log, null)
			optionWindow = ResourcesCompat.getColor(res, color.option_window, null)
			optionWindowCheckbox = ResourcesCompat.getColor(res, color.option_window_checkbox, null)
		}
	}
}

/**
 * 외부 테마 패키지: PackageManager를 통해 다른 앱의 리소스 로드
 */
class ThemeResources(
	context: Context,
	private val packageName: String,
	fullLoad: Boolean = false,
) : IThemeResources {
	var resources: Resources = try {
		context.packageManager.getResourcesForApplication(packageName)
	} catch (_: PackageManager.NameNotFoundException) {
		context.resources
	}
	var defaultRes: Resources = context.resources

	override var icon: Drawable
	override var name: String
	override var author: String
	override var version: String

	override var playbg: Drawable? = null
	override var customLogo: Drawable? = null
	override var btn: Drawable? = null
	override var btnPressed: Drawable? = null
	override var chainled: Drawable? = null
	override var chain: Drawable? = null
	override var chainSelected: Drawable? = null
	override var chainGuide: Drawable? = null
	override var phantom: Drawable? = null
	override var phantomVariant: Drawable? = null
	override var xmlPrev: Drawable? = null
	override var xmlPlay: Drawable? = null
	override var xmlPause: Drawable? = null
	override var xmlNext: Drawable? = null
	override var checkbox: Int? = null
	override var traceLog: Int? = null
	override var optionWindow: Int? = null
	override var optionWindowCheckbox: Int? = null
	override var isChainLed = true

	init {
		icon = requireNotNull(getDrawable("theme_ic", drawable.theme_ic))
		version = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
			context.packageManager.getPackageInfo(packageName, PackageManager.PackageInfoFlags.of(0)).versionName ?: "unknown"
		} else {
			@Suppress("DEPRECATION")
			context.packageManager.getPackageInfo(packageName, 0).versionName ?: "unknown"
		}
		name = requireNotNull(getString("theme_name", string.theme_name))
		author = requireNotNull(getString("theme_author", string.theme_author))
		if (fullLoad) {
			playbg = getDrawable("playbg", drawable.playbg)
			customLogo = try {
				getDrawable("custom_logo")
			} catch (_: Resources.NotFoundException) {
				null
			}
			btn = getDrawable("btn", drawable.btn)
			btnPressed = getDrawable("btn_", drawable.btn_)
			try {
				chainled = getDrawable("chainled")
			} catch (_: Resources.NotFoundException) {
				isChainLed = false
				chain = getDrawable("chain", drawable.chain)
				chainSelected = getDrawable("chain_", drawable.chain_)
				chainGuide = getDrawable("chain__", drawable.chain__)
			}
			phantom = getDrawable("phantom", drawable.phantom)
			phantomVariant = try {
				getDrawable("phantom_")
			} catch (_: Resources.NotFoundException) {
				null
			}
			xmlPrev = getDrawable("xml_prev", drawable.xml_prev)
			xmlPlay = getDrawable("xml_play", drawable.xml_play)
			xmlPause = getDrawable("xml_pause", drawable.xml_pause)
			xmlNext = getDrawable("xml_next", drawable.xml_next)
			checkbox = getColor("checkbox", color.checkbox)
			traceLog = getColor("trace_log", color.trace_log)
			optionWindow = getColor("option_window", color.option_window)
			optionWindowCheckbox =
				getColor("option_window_checkbox", color.option_window_checkbox)
		}
	}

	@SuppressLint("DiscouragedApi")
	@Throws(Resources.NotFoundException::class)
	private fun getResourceId(resources: Resources, type: String, customId: String): Int {
		return resources.getIdentifier(customId, type, packageName)
	}

	fun getDrawable(resName: String, defaultId: Int? = null): Drawable? {
		return try {
			val resId = getResourceId(resources, "drawable", resName)
			ResourcesCompat.getDrawable(resources, resId, null)
		} catch (e: Resources.NotFoundException) {
			if (defaultId != null)
				ResourcesCompat.getDrawable(defaultRes, defaultId, null)
			else
				throw e
		}
	}

	fun getColor(resName: String, defaultId: Int? = 0): Int? {
		return try {
			val resId = getResourceId(resources, "color", resName)
			ResourcesCompat.getColor(resources, resId, null)
		} catch (e: Resources.NotFoundException) {
			if (defaultId != null)
				ResourcesCompat.getColor(defaultRes, defaultId, null)
			else
				throw e
		}
	}

	fun getString(resName: String, defaultId: Int?): String? {
		return try {
			val resId = getResourceId(resources, "string", resName)
			resources.getString(resId)
		} catch (e: Resources.NotFoundException) {
			if (defaultId != null)
				defaultRes.getString(defaultId)
			else
				throw e
		}
	}
}

fun loadTheme(context: Context, themeId: String, fullLoad: Boolean = false): IThemeResources {
	return when {
		themeId.startsWith("asset://") -> {
			val themeName = themeId.removePrefix("asset://")
			AssetThemeResources(context, "themes/$themeName", fullLoad)
		}
		themeId.startsWith("zip://") -> {
			val folderName = themeId.removePrefix("zip://")
			val themesDir = context.getExternalFilesDir(null)?.let { java.io.File(it, "themes/$folderName") }
				?: throw java.io.FileNotFoundException("External files dir not available")
			ZipThemeResources(context, themesDir, fullLoad)
		}
		themeId == context.packageName || themeId == BuildConfig.APPLICATION_ID -> {
			DefaultThemeResources(context, fullLoad)
		}
		else -> {
			ThemeResources(context, themeId, fullLoad)
		}
	}
}